src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
 
     return choices[0]
 }
+
+// TODO: Cache these results? Should we have a separate store? Can we also use a store with values from the explorer tree?
+// If we move this to a cache, we should remove the awsCallLogic and instead listen/send events to/from the cache.
+// (this should also be retooled to not use the request/response objects directly)
+// TODO: Add manual pagination?
+/**
+ * Grants a `vscode.QuickPick` iteration capabilities:
+ * * Pause/unpause loading
+ * * Refresh capabilities
+ * * No item/error nodes
+ *
+ * External control is still done via `picker.promptUser` as if it were a normal QuickPick.
+ */
+export class IteratingQuickPickController<TResponse> {
+    private isDone: boolean = false
+    private isPaused: boolean = false
+    private iterator: AsyncIterator<vscode.QuickPickItem[]>
+    private activeIterationTime: Date
+
+    // Default constructs are public static so they can be validated aganist by other functions.
+    public static readonly REFRESH_BUTTON: vscode.QuickInputButton = {
+        iconPath: new vscode.ThemeIcon('refresh'),
+        tooltip: localize('AWS.generic.refresh', 'Refresh'),
+    }
+    public static readonly NO_ITEMS_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.noItemsFound.label', '[No items found]'),
+        detail: localize('AWS.picker.dynamic.noItemsFound.detail', 'Click here to go back'),
+        alwaysShow: true,
+    }
+    public static readonly ERROR_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.errorNode.label', 'There was an error retrieving more items.'),
+        alwaysShow: true,
+    }
+
+    /**
+     * @param quickPick A `vscode.QuickPick` to grant iterating capabilities
+     * @param populator An IteratingQuickPickPopulator which can call an iterator and return QuickPickItems.
+     */
+    public constructor(
+        private readonly quickPick: vscode.QuickPick<vscode.QuickPickItem>,
+        private readonly populator: IteratingQuickPickPopulator<TResponse>
+    ) {
+        // append buttons specific to iterating quickPick
+        this.quickPick.buttons = [...this.quickPick.buttons, IteratingQuickPickController.REFRESH_BUTTON]
+
+        this.iterator = this.populator.getPickIterator()
+
+        this.activeIterationTime = new Date()
+    }
+
+    /**
+     * Pauses any existing item loading.
+     * Call upon selecting an item in order to pause additional background loading.
+     */
+    public pauseRequests(): void {
+        this.isPaused = true
+    }
+
+    /**
+     * Starts item loading. Restarts from existing state if paused.
+     * Useful for manual pagination, losing quick pick focus (e.g. via external link), throttling, etc.
+     */
+    public startRequests(): void {
```

Comment:
Consider the expected behavior from someone calling this multiple times. It seems like multiple callers shouldn't be allowed to load items at the same time.

I think you can

* automatically start loading requests upon constructor
* change this method to `resumeRequests`. If `pauseRequests` was called, this would resume. If it's already loading requests, this would have no effect.

Reply:
Not sure why multiple callers would call the same instance multiple times, they should each spin up a separate instance. This prevents background loading since creating and prompting a QuickPick are two separate actions.

Reply:
> This prevents background loading since creating and prompting a QuickPick are two separate actions.

Yeah it's fair. I think it's fine either way as far as auto-starting. Common pattern I've seen is to use a constructor param to control the behavior of auto-starting. But not a big deal here.

> Not sure why multiple callers would call the same instance multiple times, they should each spin up a separate instance.

Yeah that's the point for defensive coding here. You should protect against someone calling `startRequests` multiple times either by mistake or due to a race in the calling code. You can do this here by either doing *nothing* or throwing an Error. I would opt for doing *nothing* because it's a reasonable expectation if it's already started.

It's kind of like hitting the play button on a video you're watching. With the current code, it pauses the playing video and opens a new window and starts playing the video again from the beginning. You would expect it to be a no-op.